doc/design-docs/VTGateBuffering.md

```diff
@@ -0,0 +1,60 @@
+all tables are in the process of starting up.
+
+If either of these occurs, the _consistent_ flag is set to false for that keyspace. When that happens the keyspace
+watcher checks, on every SrvKeyspace update, if the event has got resolved. This can happen when tablets are now
+available (in case of a cluster event) or if the partition information indicates that resharding is complete.
+
+When that happens. the keyspace event watcher publishes an event that the keyspace is now consistent. The buffers are
+then drained and the queries retried (at the tablet gateway).
+
+## Additional support for MoveTables
+
+### Background
+
+MoveTables does not affect the entire keyspace, just the tables being moved. So the KEW doesn't detect a cluster event
+since the tablets are still available and shard partitions are unchanged.
+
+MoveTables moves tables from one keyspace to another. There are two flavors of MoveTables: one where the tables are
+moved into all shards in the target keyspace. In Shard-By-Shard Migration user can specify a subset of shards to move
+the tables into.
+
+These are the topo attributes that are affected during a MoveTables (regular or shard-by-shard):
+
+* DeniedTables in a shard's TabletControls. These are used to stop writes to the source keyspace for these tables. While
+  switching writes we first create these entries, wait for the target to catchup to the source (using gtid positions),
+  and then update the routing rules to point these tables to the target.
+* RoutingRules (for regular movetables) and ShardRoutingRules (for shard by shard migration). Simplified, routing rules
+  are pointers for each table being moved to a keyspace. When a MoveTables is initiated, that keyspace is the source
+  keyspace. After traffic is switched the pointer is changed to point to the target keyspace. If routing rules 
+  are specified, VTGate uses them to decide which keyspace to route each table in a query to. 
+
+### Changes
+
+There are two main changes:
+* The keyspace event watcher is enhanced to look at the topo attributes mentioned above. We need to watch the 
+  SrvVSchema for changes to the Routing Rules. DeniedTables are only in the Shard records in the topo. To get around 
+  that we change the traffic switcher to also rebuild SrvVSchema when DeniedTables are modified. This way we get 
+  notified when that changes as well.
+* The logic to start buffering needs to look for the "enforce denied tables" error that is thrown by the 
+  vttablets when it tries to execute a query on a table being switched.
+* We cannot use the current buffering logic which is at the tablet gateway: meaning the keyspace is
+  already fixed by the planner and cannot be changed in that layer. We need to add a new buffering logic at a higher
+  level (the _newExecute_ method) and always replan before retrying a query. This also means that we need to bypass 
+  the plan cache while retrying.
```

Comment:
I believe buffering is still happening at tabletGateway. Only the retry is moved up to re-plan the query when buffering is stopped.